packages/hardhat-core/src/internal/hardhat-network/provider/BlockchainData.ts

```diff
@@ -14,6 +19,27 @@ export class BlockchainData {
   private _transactions: Map<string, TypedTransaction> = new Map();
   private _transactionReceipts: Map<string, RpcReceiptOutput> = new Map();
   private _totalDifficulty: Map<string, BN> = new Map();
+  public blockReservations: Array<{
+    first: BN;
+    last: BN;
+    interval: BN;
+  }> = new Array();
+
+  public reserveBlocks(first: BN, count: BN, interval: BN, common: Common) {
+    const last = first.add(count);
+    this.blockReservations.push({ first, last, interval });
+    this.addBlock(
```

Comment:
This is a confusing topic. Thank you for raising these questions.

First off, you were right about block 4 in your example; I've addressed that off-by-1 error in https://github.com/nomiclabs/hardhat/pull/2032/commits/37346fc9f8016d3dcd6ca5d6fe06d59077c18c7f.

You're also right that it's nonsensical to have that "capstone block" be replaced upon "fulfillment."  I'm going to change it so that block is not considered "reserved," so that it will not be replaced upon access.

Furthermore, after discussing with Pato, we decided that the sequence of blocks mined via `hardhat_mine` will consist of:
* regular block(s) comprised of any pending transactions
* one or more "reserved" blocks
* a "fulfilled reservation" block
* a regular (fully valid) block

The change from the previous model is the addition of the regular block at the end.  Under this new model, the final block in the mined sequence will be fully valid.  And, whereas before we were saying that you needed to request at least 2 blocks in order to induce a "reservation" (assuming that there are no pending transactions), now we're going to be saying that you need to request at least 3 blocks in order to induce a reservation.

I'll add this to the checklist: Consider the "capstone block" as a fulfilled reservation, not an unfulfilled one; and, ensure that the mined sequence of blocks is terminated by a fully valid block.

Reply:
Addressed in https://github.com/nomiclabs/hardhat/pull/2032/commits/88507135d25937d7d8a96488e7eff951cb5427ad